File: go/git-cthulhu/main_test.go
package main

import (
	"encoding/json"
	"fmt"
	"gopkg.in/yaml.v3"
	"gotest.tools/v3/assert"
	"io"
	"os"
	"testing"
)

type TestCase struct {
	Case  string      `json:"name"`
	Given []GitCommit `json:"given"`
	Want  []string    `json:"want"`
}

func TestDrawGraph(t *testing.T) {
	tests := readTestCases("./tests/DrawGraph.yaml")

	for i := range tests {
		test := &tests[i]
		test.Given = append(test.Given, GitCommit{Parents: []string{}})
		for j := range test.Given {
			commit := &test.Given[j]
			commit.Hash = fmt.Sprintf("%d", len(test.Given)-j)
		}
	}

	debug(tests)

	for _, tc := range tests {
		t.Run(tc.Case, func(t *testing.T) {
			got := DrawGraph(tc.Given)
			assert.DeepEqual(t, got, tc.Want)
		})
	}
}

func debug[T any](tests T) {
	x, err := json.MarshalIndent(tests, "", "    ")
	check(err)
	fmt.Printf("%v\n", string(x))
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func readTestCases(filePath string) []TestCase {
	var cases []TestCase
	file, err := os.Open(filePath)
	check(err)
	textBytes, err := io.ReadAll(file)
	check(err)
	yaml.Unmarshal(textBytes, &cases)
	return cases
}

File: go/git-cthulhu/prototype/main.test.ts
import { test, expect, describe } from "bun:test";
import * as yaml from "yaml";
import { DrawGraph, Git } from "./main";

const specFile = "../spec/DrawGraph.yaml";

type TestCase = {
  case: string;
  want: string[];
  given: Pick<Git.Commit, "parents">[];
};

describe("DrawGraph()", async () => {
  const specs: TestCase[] = yaml.parse(await Bun.file(specFile).text());

  for (const spec of specs) {
    test(spec.case, () => {
      expect(DrawGraph({ commits: []})).toEqual(spec.want);
    });
  }

  expect(true);
});
